new PR EIP-3475 (#3475)
ethereum/EIPs: EIPS/eip-3473.md

---
eip: 3473
title: ERC-659 Multiple Callable Bonds Standard

> author: Yohji Sakamoto <info@sgm.finance>

Reviewer: ```suggestion
author: Yohji Sakamoto (@sgmfinance)
```
We require authors to include at least one github username in the authors field because it allows that author interact with the automerge bot for update to the EIP.

discussions-to: https://github.com/ethereum/EIPs/issues/3467
type: Standards Track
category: ERC

> status: Review

Reviewer 2: ```
status: Draft
```
EIPs start in Draft, needs to merge as that first.

created: 2021-04-05
requires: 20
---

## Simple Summary

> A standard interface for contract, that manage multiple callable bonds. A single contract includes any given number of bond classes, bond nonce, bond balance of an address. This standard provides independent functions to read, transfer any collection of bonds, as well as allow bonds to be redeemed from the bond issuer if certain conditions are met. This token standard can replace current ERC20 LP token. ERC-659 has more complex data structure, which will allow the LP token to store more information, and allow the developer to build more sophisticated logic for the redemption and reward system of the DEFI project in question.

Reviewer: ```suggestion
A standard interface for contract, that manage multiple callable bonds. A single contract includes any given number of bond classes, bond nonce, bond balance of an address. This standard provides independent functions to read, transfer any collection of bonds, as well as allow bonds to be redeemed from the bond issuer if certain conditions are met. This token standard can replace current ERC-20 LP token. ERC-659 has more complex data structure, which will allow the LP token to store more information, and allow the developer to build more sophisticated logic for the redemption and reward system of the defi project in question.

```

> ![img](https://github.com/sgmfinance/erc-659/blob/main/blob/master/BOND.png?raw=true) 

Reviewer: Please bring this image into the EIPs repo. Refer to EIP-1 to see how where it should be place in the repo.

```
event eventRedeemBond(address _operator, address _from, uint256 class, uint256 nonce, uint256 _amount);
```

`"eventBurnBond"` MUST trigger when Bonds are burned. This SHOULD not include zero value burning. When redeem a bond MUST not create this event(Use`"event redeemBond" ` instead).

e.g. 

`"burn by address(_operator) 500 SASH-USD Bond(Nonce14) from 0x2d03B6C79B75eE7aB35298878D05fe36DC1fE8Ef"`

```
event eventBurnBond(address _operator, address _from, uint256 class, uint256 nonce, uint256 _amount);
```

`"eventTransferBond"` MUST trigger when Bonds are transferred. This SHOULD not include zero value transfers. Transfer event with the `_from` `0x0` MUST not create this event(Use`"event issueBond" `instead ). Transfer event with the `_to` `0x0` MUST not create this event(Use`"event redeemBond" `when redemption, and `"event burnBond" `when burning).

e.g.

`"transfer by address(_operator) 500 SASH-USD Bond(Nonce14) from 0x2d03B6C79B75eE7aB35298878D05fe36DC1fE8Ef, to address(_to)"`

```
event eventTransferBond(address _operator, address _from, address _to, uint256 class, uint256 nonce, uint256 _amount);
```

## Backwards Compatibility

> ERC-659 contract is not Compatible with contracts that don't have an ERC-659 interface built in. This requires the existing contract to upgrade with ERC-659 interface. The receiving of ERC-659 Bond need the implementation of ERC659 interface in the receiver contract.

Reviewer: ```suggestion
ERC-659 contract is not compatible with contracts that don't have an ERC-659 interface built in. This requires the existing contract to upgrade with ERC-659 interface. The receiving of ERC-659 Bond need the implementation of ERC-659 interface in the receiver contract.

```

However any existing ERC-20 token contract can issue their ERC-659 bond, by giving the minting role to a bonk contract with ERC-659 interface built in. The implementation of This can be found in our Use Cases.

To ensure the reading of transactions, `"eventIssueBond"` ,`"eventRedeemBond"` ,`"eventBurnBond"` ,`"eventTransferBond"`, Events cited above MUST be emitted when such transaction is passed. 

> ***Note that the ERC 659 interface is also compatible with ERC-20 and ERC-721 interface .*** But the creation of a separated bank contract is recommended for reading and future upgrade needs and .

Reviewer: ```suggestion
***Note that the ERC-659 interface is also compatible with ERC-20 and ERC-721 interface .*** But the creation of a separated bank contract is recommended for reading and future upgrade needs and .

```

The issuing of ERC-659 bonds is not limited to ERC-20 token. Standard like ERC-721 nonfungible token can also issue their bond with the help of ERC-659 interface. 

Any ERC659 bond or any hybrid of ERC-20 and ERC-721 contract can be used as the collateral for another ERC-659 bond. This allows the market to create bond represent no longer a single type of ERC-20 token. Bond can now represent a collection of collaterals in ERC-20 token, ERC-721 nonfungible token and ERC-659 Multiple callable bonds...

## Use Cases

https://github.com/sgmfinance/erc-659

> **This demonstration show only a simple application of ERC659 Multiple Callable Bonds Standard. Developers can build much sophisticated logic with this interface. No function used in the example serves as a guild-line.

Reviewer: ```suggestion
**This demonstration show only a simple application of ERC-659 Multiple Callable Bonds Standard. Developers can build much sophisticated logic with this interface. No function used in the example serves as a guild-line.

```

## Reference

https://eips.ethereum.org/EIPS/eip-20

https://eips.ethereum.org/EIPS/eip-721

## Security Considerations

There are no known security considerations for this EIP. More security considerations will be added after the authoring/feedback process of this EIP.

## Copyright

### License 

> ###### apache 2.0

Reviewer: EIPs must be CC0 copyright. Please use the copyright section from the EIP template.

Author: Hi all the problems are fixed, waiting for review

### License 

> ###### CC0

Reviewer: Please use the exact same text from the EIP template.

Author: hi please check the last version(e11261a), the license part is changed as it in the template

Author: suggestion applied
